Enhance enablerepo, disablerepo

sshedi opened the pull request:
Accept comma-separated repository names and globs as well.

Currently, enabling or disabling specific repositories requires multiple flags, for example:
```
tdnf ...  --disablerepo=repo1 --disablerepo=repo2 --enablerepo=repo3 --enablerepo=repo4
```

This PR extends the behavior so that each flag can take a comma-separated list, allowing the same operation with:
```
tdnf ...  --disablerepo=repo1,repo2 --enablerepo=repo3,repo4
```

Globs are also supported.

sshedi committed:
Reset nMatch at start of each iteration in TDNFAlterRepoState

Ensures match state does not carry over between repo list entries and
fixes incorrect enable/disable behavior when processing multiple repos.

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>

sshedi committed:
Support comma-separated repo names for enablerepo/disablerepo

Allow multiple repos to be specified in --enablerepo and --disablerepo
options using comma-separated values. Glob patterns work with comma-
separated values, enabling patterns like --enablerepo=photon*,test*.

Add tests for comma-separated names, globs, and mixed patterns.

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>

oliverkurth commented:
I wonder if we can use the `FNM_EXTMATCH` flag instead for `fnmatch()`, see the man page:
```
       FNM_EXTMATCH
              If this flag (a GNU extension) is set, extended patterns are supported, as introduced by 'ksh' and now supported by other shells.  The extended format is as follows, with pattern-list being a '|' separated list of patterns.
...
       '+(pattern-list)'
              The pattern matches if one or more occurrences of any of the patterns in the pattern-list match the input string.
...
```
This would accomplish the same thing, if we can do `--disablerepo=+(repo1,repo2)`.

I wish we could use patterns like `{foo,bar}`, but I do not think `fnmatch()` supports that, and I do not know what does.

sshedi commented:
> I wonder if we can use the `FNM_EXTMATCH` flag instead for `fnmatch()`, see the man page:
> 
> ```
>        FNM_EXTMATCH
>               If this flag (a GNU extension) is set, extended patterns are supported, as introduced by 'ksh' and now supported by other shells.  The extended format is as follows, with pattern-list being a '|' separated list of patterns.
> ...
>        '+(pattern-list)'
>               The pattern matches if one or more occurrences of any of the patterns in the pattern-list match the input string.
> ...
> ```
> 
> This would accomplish the same thing, if we can do `--disablerepo=+(repo1,repo2)`.
> 
> I wish we could use patterns like `{foo,bar}`, but I do not think `fnmatch()` supports that, and I do not know what does.

For TDNF use cases, `FNM_EXTMATCH` could be overkill. Regex matching is hard to remember and document. For now the existing implementation is enough I guess, using brackets and some other special characters. `TDNFIsGlob` already has some amount of this, that should suffice.

oliverkurth commented on client/repolist.c:
```diff
@@ -726,33 +726,34 @@ TDNFRepoListFinalize(
-                              pTdnf->pRepos, 0, "*");
-                BAIL_ON_TDNF_ERROR(dwError);
-                nRepoidSeen = 1;
+        int isRepoFlag = -1;
```
Nit: `isRepoFlag` is a confusing name, please rename to something like `isEnabled`.

oliverkurth commented on client/repolist.c:
```diff
@@ -851,6 +852,8 @@ TDNFAlterRepoState(
 
     for (int nMatch = 0; pRepos; pRepos = pRepos->pNext)
     {
+        nMatch = 0;
```
Oh, looks like this fixes a bug? And you can remove the `nMatch = 0` in the `for(...` just above.

sshedi replied:
Yes, first commit has the details.

oliverkurth reviewed (changes requested):
Okay, I'll accept with just a few minor changes. Thank you for this!

oliverkurth commented on client/repolist.c:
```diff
@@ -726,33 +726,34 @@ TDNFRepoListFinalize(
+              nRepoidSeen = 1;
+            }
+
+            dwError = TDNFAlterRepoState(pTdnf->pRepos, 1, cn->value);
```
I just noticed this - can we have the same enhancement here? For `--repoid=foo,bar`.

sshedi committed:
Add comma-separated list support for --repoid

Enable --repoid/--repo to accept comma-separated repo IDs, similar to
--enablerepo/--disablerepo. Includes test cases.

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>
